articles/WideStrings.md

```diff
@@ -0,0 +1,186 @@
+However this changes in DDS-XTYPES 1.2 section 7.3.1.4 to be a 16bit value.
+It is expected that most DDS implementations will switch to 16bit character storage in the future.
+ROS 2 will aim to be compatible with DDS-XTYPES 1.2 and use 16bit storage for wide characters.
+Generated code for ROS 2 messages will automatically handle the conversion when a message is serialized or deserialized.
+
+### Bounded wide strings
+
+Message definitions may restrict the maximum size of a string.
+These are referred to as bounded strings.
+Their purpose is to restrict the amount of memory used, so the bounds must be specified as units of memory.
+If a `string` field is bounded then the size is given in bytes.
+Similarly the size of a bounded `wstring` is to be specified in words.
+It is the responsibility of whoever populates a bounded `string` or `wstring` to make sure it contains whole code points only.
+Partial code points are indistinuguishable from invalid code points, so a bounded string whose last code point is incomplete is not guaranteed to be published.
+
+## Runtime impact of wide string
+
+Dealing with wide strings puts more strain on the software of a system, both in terms of speed and code size.
+UTF-8 and UTF-16 are both variable width encodings, meaning a code point can take 1 to 4 bytes depending on the encoding.
+It may take multiple code points to represent a single user perceived character.
+One of the goals of ROS 2 is to support microcontrollers that are constrained by both code size and processor speed.
+Some wide string operations like splitting a string on a user perceived character may not be possible on these devices.
+
+However, whole string equality checking is the same whether using wide strings or not.
+Further splitting a UTF-8 string on an ASCII character is identical to splitting an ASCII character on an ASCII string.
+If code on a microcontroller must do string manipluation then it could assert that a `string` only contains ASCII data by ceasing to proces a string when it encounters a byte greater than 127.
+
+## What does the API look like to a user of ROS 2?
+
+### Python 3
+
+In Python the `str` type will be used for both strings and wide strings.
+Bytes of a known encoding should be converted to a `str` using [bytes.decode](https://docs.python.org/3/library/stdtypes.html#bytes.decode) before being assigned to a field.
+
+**Example**
+
+```python
+import sys
+from time import sleep
+
+import rclpy
+from rclpy.qos import qos_profile_default
+from std_msgs.msg import WString
+
+
+if __name__ == '__main__':
+    rclpy.init(sys.argv)
+
+    node = rclpy.create_node('talker')
+
+    chatter_pub = node.create_publisher(WString, 'chatter', qos_profile_default)
+
+    msg = WString()
+    msg.data = 'Hello World' + bytes([0x00, 0x00, 0x26, 0x3A]).decode('utf-32-be')
+    print('Publishing: "{0}"'.format(msg.data))
+    chatter_pub.publish(msg)
+    node.destroy_node()
+    rclpy.shutdown()
+```
+
+### C++
```

Comment:
There needs to be a note about the need for a BOM on Windows.

Reply:
Where is the BOM needed? At the start of the source file? At the start of all string literal?

Reply:
Same as ros2/rcl_interfaces#79.

Reply:
Does 3996b8d describe it accurately?

Reply:
:+1: (I added 5588980f06a0970a5e232bec65b933505355704e for clarification).